ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -0,0 +1,567 @@
+      String containerId, Long referenceTime) {
+    if (putIndexValue.getExpiresAtMs() == -1) {
+      results.updateContainerBaseMap(serviceId, containerId, putIndexValue.getSize() * -1);
+    } else if (!isExpired(putIndexValue.getExpiresAtMs(), referenceTime)) {
+      results.updateContainerBaseMap(serviceId, containerId, putIndexValue.getSize() * -1);
+      if (results.checkIfBucketEntryExists(putIndexValue.getExpiresAtMs(), serviceId, containerId)) {
+        results.updateContainerBucket(putIndexValue.getExpiresAtMs(), serviceId, containerId,
+            putIndexValue.getSize());
+      }
+    }
+  }
+
+  private boolean isWithinWindow(long referenceTime, long window, long offSet) {
+    return referenceTime < window - offSet;
+  }
+
+  private class IndexScanner implements Runnable {
+    final long scanTimeInMs;
+    final long scanTimeWithOffsetInMs;
+    final ScanResults newScanResults;
+    final long scanWindowBoundary;
+    final long requestedScanReferenceTime;
+
+    IndexScanner(long scanTimeInMs, long requestedScanReferenceTime) {
+      this.scanTimeInMs = scanTimeInMs;
+      this.scanTimeWithOffsetInMs = scanTimeInMs - segmentScanOffset;
+      this.newScanResults = new ScanResults(bucketCount, bucketTimeSpan);
+      this.scanWindowBoundary = newScanResults.createBuckets(scanTimeInMs, segmentScanOffset);
+      this.requestedScanReferenceTime = requestedScanReferenceTime;
+    }
+
+    @Override
+    public void run() {
+      Offset checkpoint = prepareForScanning();
+      Offset previousSegmentOffset = null;
+      try {
+        for (Map.Entry<Offset, IndexSegment> indexSegmentEntry : index.indexes.entrySet()) {
+          IndexSegment segment = indexSegmentEntry.getValue();
+          if (segment.getStartOffset().compareTo(checkpoint) >= 0) {
+            break;
+          }
+          List<MessageInfo> messageInfos = new ArrayList<>();
+          segment.getEntriesSince(null, new FindEntriesCondition(Integer.MAX_VALUE),
+              messageInfos, new AtomicLong(0));
+          FileSpan currentSpan = new FileSpan(segment.getStartOffset(), segment.getStartOffset());
+          Offset backwardSearchEndOffset = previousSegmentOffset == null ?
+              segment.getStartOffset() : previousSegmentOffset;
+          FileSpan backwardSearchSpan = new FileSpan(index.indexes.firstEntry().getValue().getStartOffset(),
+              backwardSearchEndOffset);
+          for (MessageInfo messageInfo : messageInfos) {
+            IndexValue indexValue = index.findKey(messageInfo.getStoreKey(), currentSpan);
+            if (indexValue.getOffset().compareTo(checkpoint) >= 0) {
+              continue;
+            }
+            populateValidSizePerSegment(messageInfo, segment.getLastModifiedTime() * 1000, segment.getLogSegmentName(),
+                backwardSearchSpan);
+            populateValidSizePerContainer(messageInfo, backwardSearchSpan);
+
+            if (requestedScanReferenceTime != -1) {
```

Comment:
If the scan is not triggered by a request (i.e. scheduled scan or the initial scan) then there is not need to populate the requestedSegmentMap in scanResults since no one is waiting for it and no one will be reading from it.